Make Error.GoError return errors.New(e.Message)

kkaneda opened the pull request:
`GoError()` used to convert Error to a corresponding error detail type, but we no longer do. To get an error detail, new method `GetDetail()` must be used.

The commit has the following changes:
- Make all error details implement `StructuredError`.
- Add `Error.GetDetail`. The method is used instead of `GoError()` when a structured error is needed (e.g., type check).
- Use `roachpb.Error` in `range_tree.go`. `InsertRange` and `DeleteRange` need to return `roachpb.Error` since they are called from `Txn` in `Admin{Merge,Split}`.

Each error detail still implements the error interface.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4058)

<!-- Reviewable:end -->

kkaneda committed:
Make Error.GoError return errors.New(e.Message)

GoError() used to convert Error to a corresponding error detail type,
but we no longer do. To get an error detail, new method GetDetail()
must be used.

The commit has the following changes:

- Make all error details implement StructuredError.

- Add Error.GetDetail. The method is used instead of GoError() when an error detail is needed.

- Use roachpb.Error in range_tree.go. InsertRange and DeleteRange need
  to return roachpb.Error since they are called from Txn in
  Admin{Merge,Split}.

Each error detail still implements the error interface.

tamird commented:
Reviewed 27 of 27 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[roachpb/errors.go, line 58 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4058#-K9SbF2svyZ6I_pvUpqh:-K9SbF2tU5RPYXD17aHe:1619003752) ([raw file](https://github.com/cockroachdb/cockroach/blob/72d28206c48479a6c28f2759da5cbd41ad21c315/roachpb/errors.go#L58)):*
nit: wouldn't you want to do the nil check before the type assertion?

---

*[roachpb/errors.go, line 135 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4058#-K9SbrQHNK3C0XroIKJw:-K9SbrQHNK3C0XroIKJx:-1975968031) ([raw file](https://github.com/cockroachdb/cockroach/blob/72d28206c48479a6c28f2759da5cbd41ad21c315/roachpb/errors.go#L135)):*
why does this method take an argument?

---

*[roachpb/errors.go, line 183 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4058#-K9SbgqC8zrr2zQYejMJ:-K9SbgqC8zrr2zQYejMK:-1592830376) ([raw file](https://github.com/cockroachdb/cockroach/blob/72d28206c48479a6c28f2759da5cbd41ad21c315/roachpb/errors.go#L183)):*
is this really the right place for this logic? i see that this change is just fallout, but this is still rather awkward.

---

*[roachpb/errors.go, line 300 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4058#-K9Sc66eRp5xbVH6trpq:-K9Sc66eRp5xbVH6trpr:-2055542225) ([raw file](https://github.com/cockroachdb/cockroach/blob/72d28206c48479a6c28f2759da5cbd41ad21c315/roachpb/errors.go#L300)):*
every one of the implementations of `message` is a copy of the `Error` implementation. Why have both?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4058)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

---

Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[roachpb/api_test.go, line 45 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4058#-K9T5nUoAQROS4yEILPB:-K9T5nUoAQROS4yEILPC:1595579268) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1588e14b9dcf9164e1d3f7e5f5e55f1be47001b/roachpb/api_test.go#L45)):*
I assume we're 100% converted to `roachpb.Error` on code paths that care about retryability?

---

*[roachpb/errors.go, line 59 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4058#-K9T212RCg-fi4yhfMLL:-K9T212RCg-fi4yhfMLM:-326216857) ([raw file](https://github.com/cockroachdb/cockroach/blob/72d28206c48479a6c28f2759da5cbd41ad21c315/roachpb/errors.go#L59)):*
I don't think `err` can ever be nil here. The type assertion will panic if `Detail.GetValue()` returns nil.  maybe this should use a `, ok` type assertion?

---

*[roachpb/errors.go, line 132 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4058#-K9T2QPzYVhOPVrmagjR:-K9T2QPzYVhOPVrmagjS:-897930688) ([raw file](https://github.com/cockroachdb/cockroach/blob/72d28206c48479a6c28f2759da5cbd41ad21c315/roachpb/errors.go#L132)):*
`StructuredError` is an odd name for this because the structured information is not actually in the detail, it's in the `pErr` that gets passed in to `message()`.  Maybe it should be called `ErrorDetail` instead.

---

*[roachpb/errors.go, line 156 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4058#-K9T3fZtrPhvBaPWMFFV:-K9T3fZtrPhvBaPWMFFW:-1269724184) ([raw file](https://github.com/cockroachdb/cockroach/blob/72d28206c48479a6c28f2759da5cbd41ad21c315/roachpb/errors.go#L156)):*
Why does this look at `e.Detail`? Shouldn't `e.Detail` just get clobbered by the incoming `err` (if it is a detail type)?

---

*[roachpb/errors.go, line 183 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4058#-K9SbgqC8zrr2zQYejMJ:-K9T49ru8PbPXm_d_jkM:-678377371) ([raw file](https://github.com/cockroachdb/cockroach/blob/72d28206c48479a6c28f2759da5cbd41ad21c315/roachpb/errors.go#L183)):*
Again I think the type assertion should use `, ok` to guard against nil (which could happen due to version skew and other nodes that know about error codes that we don't have yet)

---

*[roachpb/errors.go, line 300 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4058#-K9Sc66eRp5xbVH6trpq:-K9T4sIsn3tCSNN3XAr_:-1622697642) ([raw file](https://github.com/cockroachdb/cockroach/blob/72d28206c48479a6c28f2759da5cbd41ad21c315/roachpb/errors.go#L300)):*
I assume all the `Error()` methods will be going away soon, but if they're going to stick around then it would be good to define them as `e.message(nil)` where possible.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4058#-:-K9T5umW-Tpcmj-ugnRc:676834233)*
<!-- Sent from Reviewable.io -->

kkaneda commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[roachpb/api_test.go, line 45 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4058#-K9T5nUoAQROS4yEILPB:-K9UoMtgex_TW34mgMPJ:1246595872) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1588e14b9dcf9164e1d3f7e5f5e55f1be47001b/roachpb/api_test.go#L45)):*
Yes, I've checked that. I don't have a very good way to write a test, but looked at every place where `GoError` is called. The method is currently called in the roachpb/server/storage/ts package in non-test code, and the number of total invocations is just 15. At about half of the places, we call `GoError` to pass an `error` to `queueImp.process`.

---

*[roachpb/errors.go, line 58 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4058#-K9SbF2svyZ6I_pvUpqh:-K9Upa6MSak4O2DYUSyc:2020285537) ([raw file](https://github.com/cockroachdb/cockroach/blob/72d28206c48479a6c28f2759da5cbd41ad21c315/roachpb/errors.go#L58)):*
Changed to use a `, ok `type assertion as suggested by Ben.

---

*[roachpb/errors.go, line 59 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4058#-K9T212RCg-fi4yhfMLL:-K9UpXjCxmrGZ4Zva4ma:503517194) ([raw file](https://github.com/cockroachdb/cockroach/blob/72d28206c48479a6c28f2759da5cbd41ad21c315/roachpb/errors.go#L59)):*
I see. Changed to use a `, ok `type assertion.

---

*[roachpb/errors.go, line 132 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4058#-K9T2QPzYVhOPVrmagjR:-K9Upz_bIpaL-NyIHYrs:1952043174) ([raw file](https://github.com/cockroachdb/cockroach/blob/72d28206c48479a6c28f2759da5cbd41ad21c315/roachpb/errors.go#L132)):*
Agreed.  We already define`ErrorDetail`in `errors.proto`, so used `ErrorDetailInterface`..

---

*[roachpb/errors.go, line 135 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4058#-K9SbrQHNK3C0XroIKJw:-K9Uq1LmKPvrCC84MDzX:-1742354116) ([raw file](https://github.com/cockroachdb/cockroach/blob/72d28206c48479a6c28f2759da5cbd41ad21c315/roachpb/errors.go#L135)):*
Yes, some impls of `message` need to print a `txn` stored in `pErr` (e.g., `TransactionAbortedError.message`).

---

*[roachpb/errors.go, line 156 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4058#-K9T3fZtrPhvBaPWMFFV:-K9UraQILeA-6QuO0kSd:1747135737) ([raw file](https://github.com/cockroachdb/cockroach/blob/72d28206c48479a6c28f2759da5cbd41ad21c315/roachpb/errors.go#L156)):*
You're totally right. Fixed.

---

*[roachpb/errors.go, line 183 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4058#-K9SbgqC8zrr2zQYejMJ:-K9UsR0Ntf33U5r3M4Wr:-1675059379) ([raw file](https://github.com/cockroachdb/cockroach/blob/72d28206c48479a6c28f2759da5cbd41ad21c315/roachpb/errors.go#L183)):*
Sure, changed to use the type assertion here as well. I agree that the code looks awkward. One approach might to unexpose the method and force `NewErrorWithTxn` instead. Added a TODO comment.

---

*[roachpb/errors.go, line 300 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4058#-K9Sc66eRp5xbVH6trpq:-K9UvMCrkpdUyRUhRcXg:451874585) ([raw file](https://github.com/cockroachdb/cockroach/blob/72d28206c48479a6c28f2759da5cbd41ad21c315/roachpb/errors.go#L300)):*
Yes, the plan is to unimplement `Error`, but it might take a while. Sure, changed them to call `e.message(nil)`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4058)*
<!-- Sent from Reviewable.io -->

kkaneda commented:
Thanks for the quick review, @tamird , @bdarnell ! Addressed the comments.

tamird commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4058)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4058#-:-K9YWtYZBaEC4rUZwWlb:676834233)*
<!-- Sent from Reviewable.io -->
